File: dastanlib/src/main/java/com/dastanapps/dastanlib/utils/ViewUtils.java
package com.dastanapps.dastanlib.utils;

import android.app.Activity;
import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Build;
import com.google.android.material.snackbar.Snackbar;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.view.ContextThemeWrapper;
import androidx.appcompat.widget.AppCompatEditText;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.widget.FrameLayout;
import android.widget.Toast;

import com.dastanapps.dastanlib.DastanApp;
import com.dastanapps.dastanlib.R;


/**
 * *@author : Dastan Iqbal
 *
 * @email : ask2iqbal@gmail.com
 */
public class ViewUtils {

    private static final String TAG = ViewUtils.class.getSimpleName();
    private static ProgressDialog progressDialog;
    private static ProgressDialog progressDialogNotCancellable;
    private static Snackbar snack = null;

    private static Toast mToast = null;

    /**
     * @param ctxt
     */
    public static void showProgressDialog(Context ctxt) {
        if (ctxt != null && (progressDialog == null || !progressDialog.isShowing())) {
            progressDialog = new ProgressDialog(ctxt);
            // progressDialog.setCancelable(false);
            progressDialog.setMessage("Loading");
            progressDialog.setIndeterminateDrawable(ViewUtils.getDrawable((Activity) ctxt, R.drawable.progressbar_circle));
            progressDialog.show();
            progressDialog.setCanceledOnTouchOutside(true);

            progressDialog.setOnDismissListener(new DialogInterface.OnDismissListener() {
                @Override
                public void onDismiss(DialogInterface dialog) {
                    hideProgressDialog();
//                    progressDialog.dismiss();
                }
            });

            progressDialog.setOnCancelListener(new DialogInterface.OnCancelListener() {
                @Override
                public void onCancel(DialogInterface dialog) {
                    progressDialog.dismiss();
                }
            });
        }
    }

    public static boolean isProgressDialog() {
        if (progressDialog != null)
            return progressDialog.isShowing();
        return false;
    }

    public static void hideProgressDialog() {
        if (progressDialog != null && progressDialog.isShowing()) {
            progressDialog.dismiss();
        }
    }

    public static void showProgressDialogNotCancellable(Context ctxt) {
        if (ctxt != null && (progressDialog == null || !progressDialog.isShowing())) {
            progressDialog = new ProgressDialog(ctxt);
            // progressDialog.setCancelable(false);
            progressDialog.setMessage("Loading");
            progressDialog.setIndeterminateDrawable(ViewUtils.getDrawable((Activity) ctxt, R.drawable.progressbar_circle));
            progressDialog.show();
            progressDialog.setCanceledOnTouchOutside(false);
            progressDialog.setCancelable(false);

            progressDialog.setOnDismissListener(new DialogInterface.OnDismissListener() {
                @Override
                public void onDismiss(DialogInterface dialog) {
                    hideProgressDialog();
//                    progressDialog.dismiss();
                }
            });

            progressDialog.setOnCancelListener(new DialogInterface.OnCancelListener() {
                @Override
                public void onCancel(DialogInterface dialog) {
                    progressDialog.dismiss();
                }
            });
        }
    }

    public static void hideProgressDialogNotCancellable() {
        if (progressDialog != null && progressDialog.isShowing()) {
            progressDialog.dismiss();
        }
    }


    public static void showToast(Context ctxt, String msg) {
        if (mToast != null)
            mToast.cancel();
        mToast = Toast.makeText(DastanApp.getInstance(), msg, Toast.LENGTH_LONG);
        mToast.show();
    }

    public static void showSnack(String msg, String actionStr, View snackView, View.OnClickListener listener) {
        //with action listener
        if (snack != null)
            snack.dismiss();
        snack = Snackbar.make(snackView, msg, Snackbar.LENGTH_LONG);
        View view = snack.getView();
        FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) view.getLayoutParams();
        params.gravity = Gravity.TOP;
        view.setLayoutParams(params);
        snack.setAction(actionStr, listener);
        snack.setDuration(Snackbar.LENGTH_LONG);
        snack.show();
    }

    public static void showSnack(String msg, String actionStr, View snackView, View.OnClickListener listener, String tagValue) {
        //with action listener
        if (snack != null)
            snack.dismiss();
        snack = Snackbar.make(snackView, msg, Snackbar.LENGTH_LONG);
        Snackbar.SnackbarLayout view = (Snackbar.SnackbarLayout) snack.getView();
        FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) view.getLayoutParams();
        params.gravity = Gravity.TOP;
        view.setLayoutParams(params);
        view.findViewById(R.id.snackbar_action).setTag(R.id.handleMultipleTap, tagValue);
        snack.setAction(actionStr, listener);
        snack.setDuration(Snackbar.LENGTH_LONG);
        snack.show();
    }

    public static void showSnackDown(String msg, String actionStr, View snackView, View.OnClickListener listener) {
        //with action listener
        Snackbar snack = Snackbar.make(snackView, msg, Snackbar.LENGTH_LONG);
        View view = snack.getView();
        FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) view.getLayoutParams();
        params.gravity = Gravity.BOTTOM;
        view.setLayoutParams(params);
        snack.setAction(actionStr, listener);
        snack.setDuration(Snackbar.LENGTH_INDEFINITE);
        snack.show();
    }

    public static void showSnack(String msg, View snackView, int gravity) {
        Snackbar snack = Snackbar.make(snackView, msg, Snackbar.LENGTH_LONG);
        View view = snack.getView();
        FrameLayout.LayoutParams params = (FrameLayout.LayoutParams) view.getLayoutParams();
        params.gravity = gravity;
        view.setLayoutParams(params);
        snack.show();
    }

    public static View inflateLayout(Context ctxt, int resId) {
        return LayoutInflater.from(ctxt).inflate(resId, null);
    }

    public static View inflateLayout(Context ctxt, int resId, ViewGroup parent) {
        return LayoutInflater.from(ctxt).inflate(resId, parent, false);
    }

    public static Dialog getDDialog(Context ctxt, int reslayout) {
        Dialog d = new Dialog(ctxt);
        d.requestWindowFeature(Window.FEATURE_NO_TITLE);
        d.getWindow().setBackgroundDrawableResource(R.color.transparent);
        d.setContentView(reslayout);
        d.setCanceledOnTouchOutside(false);
        return d;
    }

    public static Dialog getDDialogOK(Context ctxt, String title, String msg,
                                      String positiveText, String negativeText, DialogInterface.OnClickListener okInterface,
                                      DialogInterface.OnClickListener cancelInterface) {
        AlertDialog.Builder builder = new AlertDialog.Builder(ctxt);
        builder.setTitle(title);
        builder.setMessage(msg);
        builder.setPositiveButton(positiveText, okInterface);
        builder.setNegativeButton(negativeText, cancelInterface);
        AlertDialog alertDialog = builder.create();
        return alertDialog;
    }

    public static AlertDialog getDDialogOK(Context ctxt, String title, String msg,
                                           String positiveText, DialogInterface.OnClickListener okInterface, boolean isCancelable) {
        AlertDialog.Builder builder = new AlertDialog.Builder(ctxt);
        builder.setTitle(title);
        builder.setMessage(msg);
        builder.setCancelable(isCancelable);
        builder.setPositiveButton(positiveText, okInterface);

        AlertDialog alertDialog = builder.create();
        return alertDialog;
    }

    public static AlertDialog getCustomDDialogOK(Context ctxt, String title, String msg,
                                                 String positiveText, DialogInterface.OnClickListener okInterface, boolean isCancelable) {
        AlertDialog.Builder builder = new AlertDialog.Builder(new ContextThemeWrapper(ctxt, R.style.AlertDialogCustom));
        builder.setTitle(title);
        builder.setMessage(msg);
        builder.setCancelable(isCancelable);
        builder.setPositiveButton(positiveText, okInterface);
        AlertDialog alertDialog = builder.create();
        return alertDialog;
    }

    /**
     * used to support all versions of OS
     *
     * @param ctxt
     * @param res
     * @return
     */
    public static Drawable getDrawable(Activity ctxt, int res) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            return ctxt.getResources().getDrawable(res, ctxt.getTheme());
        } else {
            return ctxt.getResources().getDrawable(res);
        }
    }

    public static void changeEditTextBottomColor(AppCompatEditText tvText, int color) {
        Drawable drawable = tvText.getBackground(); // get current EditText drawable
        drawable.setColorFilter(color, PorterDuff.Mode.SRC_ATOP); // change the drawable color

        if (Build.VERSION.SDK_INT > 16) {
            tvText.setBackground(drawable); // set the new drawable to EditText
        } else {
            tvText.setBackgroundDrawable(drawable); // use setBackgroundDrawable because setBackground required API 16
        }
    }
}

File: ui/src/main/java/com/dastanapps/dastanlib/utils/ViewUtils.kt
package com.dastanapps.dastanlib.utils

import android.app.Activity
import android.app.Dialog
import android.app.ProgressDialog
import android.content.Context
import android.content.DialogInterface
import android.graphics.PorterDuff
import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.os.Build
import android.support.design.widget.Snackbar
import android.support.design.widget.TextInputLayout
import android.support.v7.app.AlertDialog
import android.support.v7.view.ContextThemeWrapper
import android.support.v7.widget.AppCompatEditText
import android.text.TextUtils
import android.view.*
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.TextView
import android.widget.Toast
import com.dastanapps.dastanlib.DastanLibApp
import com.dastanapps.dastanlib.ui.R
import com.dastanapps.dastanlib.utils.StringUtils.stringLengthShouldbeLessthan

/**
 * *@author : Dastan Iqbal
 *
 * @email : ask2iqbal@gmail.com
 */

object ViewUtils {

    private val TAG = ViewUtils::class.java.simpleName
    private var progressDialog: ProgressDialog? = null
    private val progressDialogNotCancellable: ProgressDialog? = null
    private var snack: Snackbar? = null

    private var mToast: Toast? = null

    val isProgressDialog: Boolean
        get() = if (progressDialog != null) progressDialog!!.isShowing else false

    /**
     * @param ctxt
     */

    fun showProgressDialog(ctxt: Context?) {
        if (ctxt != null && (progressDialog == null || !progressDialog!!.isShowing)) {
            progressDialog = ProgressDialog(ctxt)
            // progressDialog.setCancelable(false);
            progressDialog!!.setMessage("Loading")
            progressDialog!!.setIndeterminateDrawable(getDrawable(ctxt as Activity, R.drawable.progressbar_circle))
            progressDialog!!.show()
            progressDialog!!.setCanceledOnTouchOutside(true)

            progressDialog!!.setOnDismissListener {
                hideProgressDialog()
                //                    progressDialog.dismiss();
            }

            progressDialog!!.setOnCancelListener { progressDialog!!.dismiss() }
        }
    }

    fun hideProgressDialog() {
        if (progressDialog != null && progressDialog!!.isShowing) {
            progressDialog!!.dismiss()
        }
    }

    fun showProgressDialogNotCancellable(ctxt: Context?) {
        if (ctxt != null && (progressDialog == null || !progressDialog!!.isShowing)) {
            progressDialog = ProgressDialog(ctxt)
            // progressDialog.setCancelable(false);
            progressDialog!!.setMessage("Loading")
            progressDialog!!.setIndeterminateDrawable(getDrawable(ctxt as Activity, R.drawable.progressbar_circle))
            progressDialog!!.show()
            progressDialog!!.setCanceledOnTouchOutside(false)
            progressDialog!!.setCancelable(false)

            progressDialog!!.setOnDismissListener {
                hideProgressDialog()
                //                    progressDialog.dismiss();
            }

            progressDialog!!.setOnCancelListener { progressDialog!!.dismiss() }
        }
    }

    fun hideProgressDialogNotCancellable() {
        if (progressDialog != null && progressDialog!!.isShowing) {
            progressDialog!!.dismiss()
        }
    }


    fun showToast(ctxt: Context, msg: String) {
        if (mToast != null)
            mToast!!.cancel()
        mToast = Toast.makeText(DastanLibApp.INSTANCE, msg, Toast.LENGTH_LONG)
        mToast!!.show()
    }

    fun showSnack(msg: String, actionStr: String, snackView: View, listener: View.OnClickListener) {
        //with action listener
        if (snack != null)
            snack!!.dismiss()
        snack = Snackbar.make(snackView, msg, Snackbar.LENGTH_LONG)
        val view = snack!!.view
        val params = view.layoutParams as FrameLayout.LayoutParams
        params.gravity = Gravity.TOP
        view.layoutParams = params
        snack!!.setAction(actionStr, listener)
        snack!!.duration = Snackbar.LENGTH_LONG
        snack!!.show()
    }

    fun showSnack(msg: String, actionStr: String, snackView: View, listener: View.OnClickListener, tagValue: String) {
        //with action listener
        if (snack != null)
            snack!!.dismiss()
        snack = Snackbar.make(snackView, msg, Snackbar.LENGTH_LONG)
        val view = snack!!.view as Snackbar.SnackbarLayout
        val params = view.layoutParams as FrameLayout.LayoutParams
        params.gravity = Gravity.TOP
        view.layoutParams = params
        view.findViewById<View>(R.id.snackbar_action).setTag(R.id.handleMultipleTap, tagValue)
        snack!!.setAction(actionStr, listener)
        snack!!.duration = Snackbar.LENGTH_LONG
        snack!!.show()
    }

    fun showSnackDown(msg: String, actionStr: String, snackView: View, listener: View.OnClickListener) {
        //with action listener
        val snack = Snackbar.make(snackView, msg, Snackbar.LENGTH_LONG)
        val view = snack.view
        val params = view.layoutParams as FrameLayout.LayoutParams
        params.gravity = Gravity.BOTTOM
        view.layoutParams = params
        snack.setAction(actionStr, listener)
        snack.duration = Snackbar.LENGTH_INDEFINITE
        snack.show()
    }

    fun showSnack(msg: String, snackView: View, gravity: Int) {
        val snack = Snackbar.make(snackView, msg, Snackbar.LENGTH_LONG)
        val view = snack.view
        val params = view.layoutParams as FrameLayout.LayoutParams
        params.gravity = gravity
        view.layoutParams = params
        snack.show()
    }

    fun showSnack(msg: String, snackView: View) {
        Snackbar.make(snackView, msg, Snackbar.LENGTH_LONG).show()
    }

    fun inflateLayout(ctxt: Context, resId: Int): View {
        return LayoutInflater.from(ctxt).inflate(resId, null)
    }

    fun inflateLayout(ctxt: Context, resId: Int, parent: ViewGroup): View {
        return LayoutInflater.from(ctxt).inflate(resId, parent, false)
    }

    fun getDDialog(ctxt: Context, reslayout: Int): Dialog {
        val d = Dialog(ctxt)
        d.requestWindowFeature(Window.FEATURE_NO_TITLE)
        d.window!!.setBackgroundDrawableResource(android.R.color.transparent)
        d.setContentView(reslayout)
        d.setCanceledOnTouchOutside(false)
        return d
    }

    fun getDDialogOK(ctxt: Context, title: String, msg: String,
                     positiveText: String, negativeText: String, okInterface: DialogInterface.OnClickListener,
                     cancelInterface: DialogInterface.OnClickListener): Dialog {
        val builder = AlertDialog.Builder(ctxt)
        builder.setTitle(title)
        builder.setMessage(msg)
        builder.setPositiveButton(positiveText, okInterface)
        builder.setNegativeButton(negativeText, cancelInterface)
        return builder.create()
    }

    fun getDDialogOK(ctxt: Context, title: String, msg: String,
                     positiveText: String, okInterface: DialogInterface.OnClickListener, isCancelable: Boolean): AlertDialog {
        val builder = AlertDialog.Builder(ctxt)
        builder.setTitle(title)
        builder.setMessage(msg)
        builder.setCancelable(isCancelable)
        builder.setPositiveButton(positiveText, okInterface)

        return builder.create()
    }

    fun getCustomDDialogOK(ctxt: Context, title: String, msg: String,
                           positiveText: String, okInterface: DialogInterface.OnClickListener, isCancelable: Boolean): AlertDialog {
        val builder = AlertDialog.Builder(ContextThemeWrapper(ctxt, R.style.AlertDialogCustom))
        builder.setTitle(title)
        builder.setMessage(msg)
        builder.setCancelable(isCancelable)
        builder.setPositiveButton(positiveText, okInterface)
        return builder.create()
    }

    /**
     * used to support all versions of OS
     *
     * @param ctxt
     * @param res
     * @return
     */

    fun getDrawable(ctxt: Activity, res: Int): Drawable {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            ctxt.resources.getDrawable(res, ctxt.theme)
        } else {
            ctxt.resources.getDrawable(res)
        }
    }

    fun changeEditTextBottomColor(tvText: AppCompatEditText, color: Int) {
        val drawable = tvText.background // get current EditText drawable
        drawable.setColorFilter(color, PorterDuff.Mode.SRC_ATOP) // change the drawable color

        if (Build.VERSION.SDK_INT > 16) {
            tvText.background = drawable // set the new drawable to EditText
        } else {
            tvText.setBackgroundDrawable(drawable) // use setBackgroundDrawable because setBackground required API 16
        }
    }

    fun setFontTextViews(ctxt: Context, font: Int, vararg tv: TextView) {
        val tf = Typeface.createFromAsset(ctxt.assets,
                ctxt.resources.getString(font))
        for (tview in tv) {
            tview.typeface = tf
        }
    }

    fun getDimensFromRes(ctxt: Context, dimenId: Int): Int {
        return (ctxt.resources.getDimension(dimenId) / ctxt.resources.displayMetrics.density).toInt()
    }

    /**
     * @param editText
     * @param textInputLayout
     * @param stringName
     * @return true is the fields are entered
     */

    fun isFormValid(editText: EditText, textInputLayout: TextInputLayout, stringName: String): Boolean {
        var noBlank = editText.text.toString()
        noBlank = noBlank.replace(" ", "")

        if (TextUtils.isEmpty(noBlank)) {
            textInputLayout.error = "Please Enter $stringName"
        } else if (!stringLengthShouldbeLessthan(noBlank, 256)) {
            textInputLayout.error = "Input text should be less than 256 character"
        } else {
            textInputLayout.error = null
            return true
        }
        return false
    }
}

